security/oc_cred.c

```diff
@@ -1066,6 +1071,78 @@ oc_cred_parse_credusage(oc_string_t *credusage_string)
   }
   return credusage;
 }
+
+static bool
+oc_cred_parse_certificate(const oc_sec_cred_t *cred, mbedtls_x509_crt *crt)
+{
+  const unsigned char *cert =
+    (const unsigned char *)oc_string(cred->publicdata.data);
+  size_t cert_size = cred->publicdata.data.size;
+  if (cert == NULL) {
+    OC_ERR("failed to get validity times from cert: %s", "empty public data");
+    return false;
+  }
+
+  mbedtls_x509_crt_init(crt);
+  int ret = mbedtls_x509_crt_parse(crt, cert, cert_size);
+  if (ret < 0) {
+    OC_ERR("failed to parse certificate: %d", ret);
+    return false;
+  }
+  return true;
+}
+
+static uint64_t
+oc_cred_time_to_timestamp(mbedtls_x509_time time)
+{
+#define MONTHSPERYEAR 12 /* months per calendar year */
+  static const int days_offset[MONTHSPERYEAR] = {
+    0, 31, 59, 90, 120, 151, 181, 212, 243, 273, 304, 334
+  };
+
+  int month = time.mon - 1;
+  long year = time.year + month / MONTHSPERYEAR;
+
+  long days = (year - 1970) * 365 + days_offset[month % MONTHSPERYEAR];
+  // leap years
+  days += (year - 1968) / 4;
+  days -= (year - 1900) / 100;
+  days += (year - 1600) / 400;
+  if ((year % 4) == 0 && ((year % 100) != 0 || (year % 400) == 0) &&
+      (month % MONTHSPERYEAR) < 2) {
+    days--;
+  }
+
+  days += time.day - 1;
+  uint64_t result = days * 24;
+  result += time.hour;
+  result *= 60;
+  result += time.min;
+  result *= 60;
+  result += time.sec;
+  return result;
+}
+
+bool
+oc_cred_get_validity_period(const oc_sec_cred_t *cred, uint64_t *valid_from,
+                            uint64_t *valid_to)
+{
+  oc_assert(cred != NULL);
+  oc_assert(valid_from != NULL);
+  oc_assert(valid_to != NULL);
+
+  mbedtls_x509_crt crt;
+  if (!oc_cred_parse_certificate(cred, &crt)) {
```

Comment:
is it possible to avoid parsing(allocate) memory? Maybe this crt is already parsed in some list oc_tls.c.